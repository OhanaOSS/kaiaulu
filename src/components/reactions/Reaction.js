import React, { Component } from 'react'
import styled from "styled-components";
import {
  Row, Col,
  Card, Button, Form, InputGroup
} from 'react-bootstrap';

// enum :emotive => {heart: 0, like: 1, dislike: 2, haha: 3, wow: 4, sad: 5, angry: 6}

const CountBubble = styled.p`
  width: 1rem;
  height: 1.2rem;
  background-color: lightblue;
  border-radius: 20px;
  font-size: 0.75rem;
  text-align: center;
  vertical-align: middle;
  z-index: -1;
  margin-left: -0.25rem;
  margin-bottom: 0;
`
const Emoji = styled.div`
  cursor: pointer;
  font-size: 1.5rem;
  z-index: 1;
  right: 1px;
  position: relative;
`
const EmojiContainer = styled(Form)`
  display: flex;
  flex-direction: row;
`

class Count extends Component {
  constructor(props){
    super(props)
    this.state = {
      count: 0
    }
  }

  componentDidMount(){
    this.setState({count: this.props.count})
  }

  componentWillReceiveProps(nextProps){
    if (nextProps.count !== this.state.count) {
      this.setState({count: nextProps.count})
    }
  }

  render() {
    if (this.state.count > 0) {
      console.log("count",this.state.count)
      return <CountBubble>{this.state.count}</CountBubble>
    } else {
      return <></>
    }
  }
}

class Reaction extends Component {
  constructor(props){
    super(props)
    this.state = {
      enabled: false,
      count: 0
    }
    this.handleClick = this.handleClick.bind(this)
  }
  handleClick(e){
    this.props.handleRequest({
      type: e.target.getAttribute('name'), 
      enabled: this.state.enabled
    })
  }
  componentWillReceiveProps(nextProps) {
    if(nextProps.count !== this.state.count) {
      this.setState({count: nextProps.count})
    }
    if(nextProps.enabled !== this.state.enabled) {
      this.setState({enabled: nextProps.enabled})
    }
  }
}


export class Heart extends Reaction {
  render() {
    return (
      <EmojiContainer>
        <Emoji onClick={this.handleClick} name={"heart"}>❤️</Emoji><Count count={this.state.count}/>
      </EmojiContainer>
      )
  }
}

export class Like extends Reaction {
  render() {
    return (
      <EmojiContainer>
        <Emoji onClick={this.handleClick} name={"like"}>👍</Emoji><Count count={this.state.count}/>
      </EmojiContainer>
      )
  }
}

export class Dislike extends Reaction {
  render() {
    return (
      <EmojiContainer>
        <Emoji onClick={this.handleClick} name={"dislike"}>👎</Emoji><Count count={this.state.count}/>
      </EmojiContainer>
      )
  }
}

export class Haha extends Reaction {
  render() {
    return (
      <EmojiContainer>
        <Emoji onClick={this.handleClick} name={"haha"}>🤣</Emoji><Count count={this.state.count}/>
      </EmojiContainer>
      )
  }
}

export class Wow extends Reaction {
  render() {
    console.log("wow state count",this.state.count)
    return (
      <EmojiContainer>
        <Emoji onClick={this.handleClick} name={"wow"}>😮</Emoji><Count count={this.state.count}/>
      </EmojiContainer>
      )
  }
}

export class Sad extends Reaction {
  render() {
    return (
      <EmojiContainer>
        <Emoji onClick={this.handleClick} name={"sad"}>😢</Emoji><Count count={this.state.count}/>
      </EmojiContainer>
      )
  }
}

export class Angry extends Reaction {
  render() {
    return (
      <EmojiContainer>
        <Emoji onClick={this.handleClick} name={"angry"}>😠</Emoji><Count count={this.state.count}/>
      </EmojiContainer>
      )
  }
}
