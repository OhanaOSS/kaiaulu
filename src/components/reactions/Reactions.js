// Note: The backend rails model has a callback after new interactions called
// "delete_old_reaction_on_interaction" making cleanup via a delete unrequired.

import React, { Component } from 'react'
import {Reaction} from './Reaction';
import {contentFetcher, contentPoster, urlBuilder} from '../../utils/contentHelper'
import {capitalizeFirstLetter} from '../../utils/dataHelpers'
import {Button} from 'react-bootstrap';
import styled from 'styled-components';
import store from '../../store';

const ReactionsWrapper = styled.ul`
  margin-top: .25rem;
  margin-bottom: .25rem;
  list-style-type: none;
  display: flex;
  flex-direction: row;
  padding-inline-start: 0px;
  li {
    margin-right: 0.5rem;
  }
`


export default class Reactions extends Component {

  constructor(props){
    super(props)
    this.state = {
      data: [],
      emotives: {
        heart: {
          count: 0, 
          active: null
        },
        like: {
          count: 0, 
          active: null
        },
        dislike: {
          count: 0, 
          active: null
        },
        haha: {
          count: 0, 
          active: null
        },
        wow: {
          count: 0, 
          active: null
        },
        sad: {
          count: 0, 
          active: null
        },
        angry: {
          count: 0, 
          active: null
        }
      },
      currentUserEmoji: null
    }
    this.handleRequest = this.handleRequest.bind(this)
    this.emojiMapper = this.emojiMapper.bind(this)
    this.checkForActiveEmoji = this.checkForActiveEmoji.bind(this)
    this.createSelectedEmoji = this.createSelectedEmoji.bind(this)
    this.deleteActiveEmoji = this.deleteActiveEmoji.bind(this)
  }
  
  componentDidMount() {
    const url = urlBuilder({
      parent_id: this.props.id,
      parent_type: this.props.type,
      request_type: "reaction"
    })
    contentFetcher(url).then(reactions => this.emojiMapper(reactions)).then(processedReactions => {
    this.setState({
      data: processedReactions.data,
      emotives: processedReactions.emotives,
      currentUserEmoji: processedReactions.currentUserEmoji
    })
    })
  }

  async handleRequest(callback) {
    this.checkForActiveEmoji(callback)
  }

  checkForActiveEmoji(clickCallback){
    let newEmotives = this.state.emotives
    Object.keys(this.state.emotives).map((key) => {
      if (this.state.emotives[key].active && key !== clickCallback.type){
        newEmotives[key].active = null
        newEmotives[key].count -= 1
      }
    })
    if (!newEmotives[clickCallback.type].active) {
      newEmotives[clickCallback.type].active = true
      newEmotives[clickCallback.type].count += 1
      this.createSelectedEmoji(clickCallback)
    } else if (newEmotives[clickCallback.type].active) {
      newEmotives[clickCallback.type].active = null
      newEmotives[clickCallback.type].count -= 1
      this.deleteActiveEmoji()
    }

    this.setState({
      emotives: newEmotives
    })
  }

  createSelectedEmoji(callback){
    const url = urlBuilder({
      parent_type: "reactions"
    })
    let data = {
      "reaction":{
        "attributes":{
          "member_id": store.getState().currentUser.data.id,
          "emotive": callback.type,
          "interaction_type": capitalizeFirstLetter(this.props.type),
          "interaction_id": this.props.id
        }
      }
    }
    contentPoster("post", data, url).then(newReaction => this.setState({currentUserEmoji: newReaction.id}))
  }
  deleteActiveEmoji(){
    const url = urlBuilder({
      parent_type: "reactions",
      parent_id: this.state.currentUserEmoji
    })
    contentPoster("delete", {}, url).then(this.setState({currentUserEmoji: null}))
  }

  emojiMapper(data){
    var emotives = {
        heart: {
          count: 0, 
          active: null
        },
        like: {
          count: 0, 
          active: null
        },
        dislike: {
          count: 0, 
          active: null
        },
        haha: {
          count: 0, 
          active: null
        },
        wow: {
          count: 0, 
          active: null
        },
        sad: {
          count: 0, 
          active: null
        },
        angry: {
          count: 0, 
          active: null
        }
      }
    let currentUserEmoji = null
    data.map(emoji => {
      var emotive = emoji.attributes["emotive"]
      var memberID = emoji.attributes["member-id"]
      var currentMemberID = store.getState().currentUser.data.id
      emotives[emotive].count += 1
      if (currentMemberID === memberID){
        currentUserEmoji = emoji.id
        emotives[emotive].active = true
      }
    })
    return {
      data: data,
      emotives: emotives,
      currentUserEmoji: currentUserEmoji
    }
  }
  shouldComponentUpdate(nextProps, nextState){
    if(nextState !== this.state || nextProps !== this.props){
      return true
    } else {
      return false
    }
  }

  render() {
    const reactions = Object.keys(this.state.emotives).map((key, index) => (
      <li key={index} as={Button}>
        <Reaction active={this.state.emotives[key].active} type={key} count={this.state.emotives[key].count} handleRequest={this.handleRequest}/>
      </li>
    ))
    return (
      <ReactionsWrapper>
        {reactions}
      </ReactionsWrapper>
    )
  }
}

