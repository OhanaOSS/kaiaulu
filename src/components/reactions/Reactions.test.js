import React from 'react';
import { shallow, mount } from 'enzyme';
import Reactions from './Reactions';
import {Reaction, Heart, Like, Dislike, Haha, Wow, Sad, Angry, Count} from './Reaction';

describe('Reactions', () => {
  it('renders properly', () => {
    const component = shallow(<Reactions type={"comments"} id={1}/>);
  
    expect(component).toMatchSnapshot();
  });
});


describe('activeEmojis being switched', () => {
  it('switch from heart(1), like(0) to heart(0), like(1)', () => {
    let component = mount(<Reactions type={"comments"} id={1}/>);
    const heartWrapper = component.find('form.reaction').at(0).children().at(0);
    const likeWrapper = component.find('form.reaction').at(1).children().at(0);
    heartWrapper.simulate('click')
    likeWrapper.simulate('click')
    let heartCounter = component.find('form.reaction').at(0).children().at(1).props().count
    let likeCounter = component.find('form.reaction').at(1).children().at(1).props().count
    // console.log(heartWrapper.debug())
    // console.log(likeWrapper.debug())
    // console.log(heartCounter)
    // console.log(likeCounter)
    expect(likeCounter).toBe(1)
    expect(heartCounter).toBe(0)
  });
});

describe('activeEmoji being added and removed', () => {
  it('switch from wow(0) => wow(1)', () => {
    const component = mount(<Reactions type={"comments"} id={1}/>);
    const reactionWrapper = component.find('form.reaction').children().first();
    let counter = component.find('Count').first().props().count;
    
    
    expect(counter).toBe(0)
    reactionWrapper.simulate('click')
    counter = component.find('Count').first().props().count;
    expect(counter).toBe(1)
  });
  it('switch from wow(1) => wow(0)', () => {
    const component = mount(<Reactions type={"comments"} id={1}/>);
    const reactionWrapper = component.find('form.reaction').children().first();
    reactionWrapper.simulate('click')
    let counter = component.find('Count').first().props().count;
    
    
    expect(counter).toBe(1)
    reactionWrapper.simulate('click')
    counter = component.find('Count').first().props().count;
    expect(counter).toBe(0)
  });
});