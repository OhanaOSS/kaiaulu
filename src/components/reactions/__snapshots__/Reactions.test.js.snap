// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reactions renders properly 1`] = `
<styled.ul>
  <li
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "propTypes": undefined,
        "render": [Function],
      }
    }
  >
    <Heart
      count={0}
      enabled={false}
      handleRequest={[Function]}
    />
  </li>
  <li
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "propTypes": undefined,
        "render": [Function],
      }
    }
  >
    <Like
      count={0}
      enabled={false}
      handleRequest={[Function]}
    />
  </li>
  <li
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "propTypes": undefined,
        "render": [Function],
      }
    }
  >
    <Dislike
      count={0}
      enabled={false}
      handleRequest={[Function]}
    />
  </li>
  <li
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "propTypes": undefined,
        "render": [Function],
      }
    }
  >
    <Haha
      count={0}
      enabled={false}
      handleRequest={[Function]}
    />
  </li>
  <li
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "propTypes": undefined,
        "render": [Function],
      }
    }
  >
    <Wow
      count={0}
      enabled={false}
      handleRequest={[Function]}
    />
  </li>
  <li
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "propTypes": undefined,
        "render": [Function],
      }
    }
  >
    <Sad
      count={0}
      enabled={false}
      handleRequest={[Function]}
    />
  </li>
  <li
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "propTypes": undefined,
        "render": [Function],
      }
    }
  >
    <Angry
      count={0}
      enabled={false}
      handleRequest={[Function]}
    />
  </li>
</styled.ul>
`;
